2단계 - 수강신청(도메인 모델)

sunggyupaik opened the pull request:
# 학습 관리 시스템(Learning Management System)

## 수강 신청 기능 요구사항
- 과정(Course)은 기수 단위로 운영하며, 여러 개의 강의(Session)를 가질 수 있다.
- 강의는 시작일과 종료일을 가진다.
- 강의는 강의 커버 이미지 정보를 가진다.
    - 이미지 크기는 1MB 이하여야 한다.
    - 이미지 타입은 gif, jpg(jpeg 포함),, png, svg만 허용한다.
    - 이미지의 width는 300픽셀, height는 200픽셀 이상이어야 하며, width와 height의 비율은 3:2여야 한다.
- 강의는 무료 강의와 유료 강의로 나뉜다.
    - 무료 강의는 최대 수강 인원 제한이 없다.
    - 유료 강의는 강의 최대 수강 인원을 초과할 수 없다.
    - 유료 강의는 수강생이 결제한 금액과 수강료가 일치할 때 수강 신청이 가능하다.
- 강의 상태는 준비중, 모집중, 종료 3가지 상태를 가진다.
- 강의 수강신청은 강의 상태가 모집중일 때만 가능하다.
- 유료 강의의 경우 결제는 이미 완료한 것으로 가정하고 이후 과정을 구현한다.
- 결제를 완료한 결제 정보는 payments 모듈을 통해 관리되며, 결제 정보는 Payment 객체에 담겨 반한된다.

## 프로그래밍 요구사항
- DB 테이블 설계 없이 도메인 모델부터 구현한다.
- 도메인 모델은 TDD로 구현한다.
    - 단, Service 클래스는 단위 테스트가 없어도 된다.
- 다음 동영상을 참고해 DB 테이블보다 도메인 모델을 먼저 설계하고 구현한다.

## 기능 목록정리
- [x] domain
  - [x] Session 생성
  - [x] Duration 생성
  - [x] Image 생성
  - [x] Type enum 생성
 - [x] service
   - [x] SessionService 생성

sunggyupaik committed:
기능 요구사항 정리

기능 요구사항을 정리합니다

sunggyupaik committed:
[feat] 강의와 이미지 도메인을 생성하라

강의와 이미지 도메인을 생성합니다. 유효성 검증만 도메인 데이터 변경
함수를 추가해야 합니다

sunggyupaik committed:
[feat] 이미지정보 유효성을 추가하라

이미지의 크기, 확장자, 비율을 검사하는 유효성을 추가합니다

sunggyupaik committed:
[feat] 강의에 유효성 검증을 추가하라

모집중인 경우만 강의신청이 가능하며, 유료 강의의 경우 수강 인원을 초과
할 수 없고 결제가 완료되어야만 신청이 가능합니다

sunggyupaik committed:
[test] 강의와 이미지 테스트를 추가하라

강의와 이미지의 도메인 유효성 테스트를 추가합니다

sunggyupaik committed:
[refactor] 강의의 필드를 개선하라

강의는 여러명의 수강이 있을 수 있으므로 수강생 리스트를 추가하고
결제정보도 여러개가 될 수 있기 때문에 결제정보 필드를 제거합니다

sunggyupaik committed:
[feat] 강의 신청 서비스 레이어를 생성하라

강의 신청 서비스레이어를 생성합니다. 강의 유효성 검사는 강의 클래스
테스트에서 진행했으므로 정상적인 강의 신청만 테스트합니다

sunggyupaik committed:
[refactor] 도메인 유효성 검증을 메서드로 분리하라

이미지, 강의 도메인 생성자의 유효성 검사를 메서드로 분리합니다

sunggyupaik committed:
[feat] 강의 신청자를 일급 컬렉션으로 감싸라

강의 신청자들을 일급 컬렉션으로 감싸도록 개선합니다

sunggyupaik committed:
[test] 이미지 타입 조회 테스트를 추가하라

이미지 타입 조회 시, 존재하지 않는 타입의 경우
예외를 던지는 테스트를 추가합니다

sunggyupaik committed:
[feat] 강의 시작과 종료날짜를 클래스로 분리하라

강의 시작과 종료날짜를 클래스로 분리하여 강의기간 생성과 유효성 검사를
책임질 수 있도록 합니다

sunggyupaik commented:
안녕하세요, step2 완성했습니다 :) 이번에도 잘 부탁드립니다.

과제하면서, 궁금한게 있습니다. qna, course 같이 분리되어 있는 도메인에서 서로의 도메인을 참조하려면 보통 id를 가지면 되나요, 아니면 객체를 가지면 되나요? `Payment`의 경우에는 `Session`과 `NsUser`를 `Long sessionId`, `Long nsUserId` 처럼 참조하는데, `Question` 과 `Answer` 에서는 id를 참조하지 않고 `NsUser writer` 같이 객체 자체를 참조하더라구요. 어떤 방식을 사용하는게 좀 더 구조적으로 좋을까요?

sunggyupaik committed:
[style] 강의의 enum 위치를 조정하라

변수 바로 아래에 내부 enum을 위치시키도록 조정합니다

ohtaeg commented on src/main/java/nextstep/courses/domain/course/Course.java:
```diff
@@ -9,6 +11,8 @@ public class Course {
 
     private Long creatorId;
 
+    private Sessions sessions;
+
     private LocalDateTime createdAt;
 
     private LocalDateTime updatedAt;
```
현재 Course가 많은 필드를 가지고 있는데요 :)
이때까지의 미션들을 진행하시면서 다음과 같은 프로그래밍 요구사항을 보셨을텐데요!

> 규칙 7: 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.

해당 객체에도 이 요구사항을 지켜볼 수 있을까요 :) ?
id, createdAt, updatedAt, creatorId 등은 다른 entity에서도 사용할 수 있는 속성일 수도 있는 점을 고려해보시면 좋을 것 같습니다 😄

sunggyupaik replied:
관련해서 맨 아래 질문 드렸습니다!

ohtaeg commented on src/main/java/nextstep/courses/domain/course/Course.java:
```diff
@@ -9,6 +11,8 @@ public class Course {
 
     private Long creatorId;
 
+    private Sessions sessions;
```
현재 해당 sessions를 사용하고 있지 않네요 😭 
TDD 과정에 새로운 강의가 추가되는 경우를 고려해볼 수 있을까요?

sunggyupaik replied:
네, 일단 간단하게 Course의 id와 Session을 통해 추가하는 서비스를 추가했습니다.!

ohtaeg commented on src/main/java/nextstep/courses/domain/course/image/Image.java:
```diff
@@ -0,0 +1,66 @@
+
+import java.time.LocalDateTime;
+
+public class Image {
```
👍  👍

ohtaeg commented on src/main/java/nextstep/courses/domain/course/session/Applicants.java:
```diff
@@ -0,0 +1,32 @@
+        return this.applicants.size();
+    }
+
+    public void add(NsUser applicant) {
+        this.applicants.add(applicant);
```
같은 학생이 중복으로 등록이 가능한 구조인 것 같아요..!

sunggyupaik replied:
네 해당 메서드에 유효성 검사 추가했습니다!

ohtaeg commented on src/main/java/nextstep/courses/domain/course/session/Session.java:
```diff
@@ -0,0 +1,130 @@
+
+    private LocalDateTime updatedAt;
+
+    public enum Type {
+        FREE("무료"),
+        CHARGE("유료");
+
+        private final String description;
+
+        Type(String description) {
+            this.description = description;
+        }
+    }
+
+    public enum Status {
```
Session에 대한 종속적인 데이터로 inner로 잘 선언해주셨네요 👍 
그렇다면 다른곳에서 접근하지 못하게 private로 선언해주시면 좋을 것 같아요 :)

sunggyupaik replied:
앗, 이부분은 한번 논의해보고 싶습니다
public으로 선언되었기 때문에 `Session.Status.RECRUIT`처럼 손쉽게 enum값을 사용 할 수 있습니다.
이는 특히 테스트에서 유용하게 사용 가능한데 이를 private로 바꿨을 경우 `find()`, `get()`등을 통해서 꺼내야 합니다.
메서드가 추가된다면 차라리 enum을 별도의 클래스로 분리하는게 좀 더 깔끔해지는 구조가 될 것 같습니다.

특정 상태를 주입하고 싶어서 원하는 Status를 얻으려면 Session에서 이런 식으로 꺼내서 사용해야 할 것 같습니다.
```java
public static Status ready() {
    return Status.READY;
}
```

public으로 남기는 허용이 코드의 일관성을 헤칠까요?

ohtaeg replied:
아하 의견 공유 감사드립니다!! 어떤 의도로 작성하셨는지 알게되었습니다!! 👍 
이 부분은 정답이 없는 부분같아 성규님이 원하시는 방향으로 하시면 좋을 것 같아요 :)

말씀해주신대로 접근성, 사용성을 우선순위로 높게 두신다면 public으로 해서 관리하는 방향도 좋다고 생각합니다! 👍 👍 

저는 캡슐화 관점에서 한번 생각해보았는데요 :)
저는 클래스 내에 이너 클래스, 이너 enum을 하는 경우는 두 클래스가 서로 긴밀한 관계이며
클래스안에서만 사용하기 위한 클래스로 사용하며 캡슐화를 통해 외부로부터 숨기는 목적으로 사용하곤 하는데요!

HashMap 내부의 Node 클래스도 외부에서 접근할 수 없도록 접근 제어를 최소화 (default)한 것 처럼 
외부로부터 숨기면 어떨까 싶어서 드렸던 코멘트였습니다 🙏 

public으로 해도 괜찮다고 생각하며 코드의 일관성도 괜찮다고 생각합니다 😄 👍

sunggyupaik replied:
네 캡슐화 관점에서 생각했을 때 private 이 맞다고 생각합니다! 단순한 로직으로 생각하고 내부 클래스로 만들었는데 다양한 로직을 추가하다보니 많이 쓰여서 별도의 enum 클래스를 가지는 것도 좋을 것 같네요 ㅎㅎ

ohtaeg commented on src/main/java/nextstep/courses/domain/course/session/Session.java:
```diff
@@ -0,0 +1,130 @@
+
+    private Image image;
+
+    private Duration duration;
```
시작일과 종료일을 하나의 객체로 다뤄주셨네요 👍 👍

ohtaeg commented on src/main/java/nextstep/courses/domain/course/session/Session.java:
```diff
@@ -0,0 +1,130 @@
+        checkStatusOnRecruit();
+
+        if (typeCharged()) {
+            checkApplySizeIsValid();
```
이건 students에게 물어보면 어떨까요~?

sunggyupaik replied:
정원 수와 수강생을 `Applicatns` 에서 관리하도록 하여 유료 강의의 경우, 정원 수를 초과하는지 Applicants에서 검사하도록 수정했습니다!

ohtaeg commented on src/main/java/nextstep/courses/domain/course/session/Session.java:
```diff
@@ -0,0 +1,130 @@
+import java.time.LocalDateTime;
+import java.util.Objects;
+
+public class Session {
```
현재 기능 요구사항에는 자세하게 안나와있지만 제 생각에는 강의 상태 변경 등의 기능도 필요하지 않을까 싶은데요 :)
기존 요구사항을 통해 새로운 요구사항이 도출된다고 생각됩니다..!

- 준비중, 모집중은 강의 시작일 전이어야 한다.
- 종료는 강의 종료일 이후여야 가능하다 등

현재 코드에서는 Duration과 Status와 연관성이 없지만 이 둘은 밀접한 관계가 있다고 생각이 되어요.
Duration와 Status를 이용하여 변경 관련 검증 로직을 추가해보시면 어떨까요~?

sunggyupaik replied:
네, 맞네요 기존 요구사항에서 도출되는 다양한 상황들을 고민해보고 필요한 내용은 구현이 필요한 것 같습니다!

ohtaeg commented on src/main/java/nextstep/courses/domain/course/session/Sessions.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public Iterator<Session> iterator() {
+        return null;
```
null을 반환하는건 위험하지 않을까요~? 👀

sunggyupaik replied:
누락되었네요, 수정했습니다

ohtaeg commented on src/main/java/nextstep/courses/infrastructure/JdbcSessionRepository.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.Optional;
+
+@Repository("sessionRepository")
+public class JdbcSessionRepository implements SessionRepository {
```
2단계 미션은 먼저 도메인 모델을 작성하는 것이 목표인데 미리 레포까지 고려해주셨네요 😄

ohtaeg commented on src/test/java/nextstep/courses/domain/course/image/ImageTest.java:
```diff
@@ -0,0 +1,47 @@
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class ImageTest {
```
👍 👍

ohtaeg commented on src/main/java/nextstep/courses/domain/course/session/Session.java:
```diff
@@ -0,0 +1,130 @@
+        }
+    }
+
+    private void checkPaymentIsPaid(NsUser loginUser, Payment payment) {
```
해당 요구사항이 누락된 것 같아요 👀 한번 확인해주시면 감사하겠습니다 :)

> 유료 강의는 수강생이 결제한 금액과 수강료가 일치할 때

sunggyupaik replied:
네, 사용자와 강의가 같으면 지불 금액까지 이미 같을거라고 생각을 했네요,
유효성과 테스트 추가헀습니다!

ohtaeg commented on src/main/java/nextstep/courses/domain/course/session/Session.java:
```diff
@@ -0,0 +1,130 @@
+import java.time.LocalDateTime;
+import java.util.Objects;
+
+public class Session {
```
Session 클래스도 많은 필드를 가지고 있는 것 같아요..! 연관된 필드끼리 묶어서 객체를 만들어 책임을 분산시켜볼 수 있을까요 :) ?

sunggyupaik replied:
관련해서 맨 아래 질문 드렸습니다!

ohtaeg commented on README.md:
```diff
@@ -1,17 +1,23 @@
-한다.
-- 질문자와 답변자가 다른경우 답변을 삭제할 수 없다.
-- 질문과 답변 삭제 이력에 대한 정보를 DeleteHistory를 활용해 남긴다.
+## 수강 신청 기능 요구사항
```
> 과제하면서, 궁금한게 있습니다. qna, course 같이 분리되어 있는 도메인에서 서로의 도메인을 참조하려면 보통 id를 가지면 되나요, 아니면 객체를 가지면 되나요? Payment의 경우에는 Session과 NsUser를 Long sessionId, Long nsUserId' 처럼 참조하는데, Question과 Answer에서는 id를 참조하지 않고 NsUser writer` 같이 객체 자체를 참조하더라구요. 어떤 방식을 사용하는게 좀 더 구조적으로 좋을까요?

이 부분은 정해진 답이 없기에 참고만 해주시면 감사하겠습니다 🙏 
어떤 도메인이냐에 따라 다를 것 같네요! 
예로 회원이라는 도메인은 왠만한 도메인에서 사용되는 공통되는 도메인으로 사용되는 경우가 많다고 생각해요
가령 주문과도 연결되고 결제와도 연결되고 상점과도 연결되고 장바구니에도 연결되는 등..

이런 경우에는 회원 도메인과의 의존성을 최소화하여 회원 도메인의 복잡도를 작게 관리하기 위해
다른 도메인들과 의존성이 많고 복잡해지는 경우 저는 간접 참조인 id를 가지도록 하는 편입니다 :)
위와같이 Question과 Answer등 의존성 방향이 명확하고 단순한 경우 바로 객체로 참조하는 편입니다..!

객체간의 관계가 단방향인지, 해당 객체 참조가 반드시 필요한 것인지 고민 해보신 후
필요 없다면 객체간 직접 참조가 아닌 id를 이용한 간접 참조도 고민해보시면 좋을 것 같아요 :)

도움이 되는 답변이였으면 좋겠네요 🙏

sunggyupaik replied:
정성스런 답변 감사합니다🙏
어떤 방식으로 고민하고 바라봐야 하는지 많은 도움 됐습니다 앞으로 다양한 상황을 겪어보고 익혀나가겠습니다~

ohtaeg commented on src/test/java/nextstep/courses/domain/course/session/SessionTest.java:
```diff
@@ -0,0 +1,103 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class SessionTest {
```
커밋들을 살펴보았는데 TDD를 잘해주신 것 같네요 💯

ohtaeg reviewed (changes requested):
2단계 구현을 잘해주셨네요 👍 👍 
다음 미션을 진행하기전에 리팩토링되면 좋을 것 같아 R.C 하였습니다..!
몇 개의 피드백들을 드렸는데 확인 후 재요청해주시면 감사하겠습니다 🙏

sunggyupaik committed:
[feat] 과정에 강의 추가를 추가하라

과정에 강의가 새롭게 추가되므로 강의를 추가하는 로직을 만듭니다

sunggyupaik committed:
[feat] 중복 신청 유효성 검사를 추가하라

과정에 강의가 중복 추가되거나, 강의에 수강생이 중복 신청되는 경우
이미 추가가 되었다는 예외를 추가합니다

sunggyupaik committed:
[feat] 강의 금액과 지불내역이 다른 유효성을 추가하라

강의 금액과 지불내역이 다르면 수강신청을 할 수 없습니다.
해당 유효성과 테스트를 추가합니다

sunggyupaik committed:
[feat] 강의 도메인 수강상태 변경 유효성을 추가하라

강의 준비와 모집은 시작일 전에 가능하며, 강의 종료는 종료일 이후에
가능하다는 유효성을 추가합니다

sunggyupaik committed:
[feat] 수강생 클래스에 정원 수를 추가하라

수강생 클래스에 단지 수강생 뿐 아니라 정원 수를 같이 관리하도록
개선합니다. 따라서 지원에 따른 정원 검사를 해당 클래스에서 책임을
가지도록 합니다

sunggyupaik committed:
[feat] 강의 상태 변경을 서비스 레이어에 추가하라

강의 준비중, 모집중, 종료 3가지 각가 상태로 변경 가능한 기능을 서비스
레이어에 추가합니다

sunggyupaik commented:
피드백 기반으로 대부분 수정했습니다.
아직 도메인 분리를 못해서 궁금한 점 공유 후에 다시 수정하겠습니다!

`Session`과 `Course`를 좀 더 작은 단위로 나누는 것에 고민이 있는데요


```java
private Long id;

private String title;

private Long creatorId;

private Sessions sessions;

//추상화 예정
//private LocalDateTime createdAt;

//private LocalDateTime updatedAt;
```

위의 변수는 `Course`의 변수입니다.  모든 클래스의 공통 부분을 분리해본다면 `id`, `createdAt`, `updatedAt` 3개 정도가 될 것 같은데요, 

아하, 글을 쓰면서 생각해봤는데 생성자, 수정자, 생성일시, 수정일시처럼 공통된 부분을 말씀해주신게 아닌가 싶네요 

그렇다면 고유 id, 생성자, 수정자, 생성일시, 수정일시 5개 칼럼을 BaseEntity 형식으로 빼서 활용하는 방법도 괜찮을까요? 찾아보니 보통 생성자, 수정자를 BaseEntity에 담고 생성일시, 수정일시를 BaseTimeEntity에 담는 것 같네요.(이번 프로젝트에서는 한 클래스에 넣겠습니다.. ㅎㅎ)  

혹시 자신의 고유 id의 경우도 따로 분리하기도 하나요? 여태까지 봤던 기억은 없어서요

sunggyupaik commented:
```java
private Long id;

private Image image;

private Duration duration;

private Type type;

private Long amount;

private Applicants applicants; //int quota 포함

private Status status;

//추상화 예정 
//private LocalDateTime createdAt; 

//private LocalDateTime updatedAt;
```
현재 Session은 위와 같이 많은 변수를 가지고 있고 그룹을 묶어보려고 합니다.

### 1. 고유 속성 도메인으로 분리하기
`duration`, `type`, `status`, `amount`는 순수하게 Session이 가지고 있는 정보니까 묶을 수 있지 않을까 생각을 했습니다.

장점은 메인 도메인인 Session이 기존 7개에서 id, image, sessionInformation, applicants 4개만 변수로 가집니다.
하지만 결국 새로운 도메인 sessionInformation도 4개의 변수를 가집니다. 덩치를 일부 분리했지만 여전히 3개 이상의 많은 변수를 담고 있습니다.

### 2. 유료/무료 강의 별로 도메인 분리하기
추상 클래스로 Session 골격을 잡아 유료와 무료 크게 구분 짓는 3가지  `type`, `amount`, `quota` 를 각 클래스에서 상속 받도록 합니다.
`ex)무료 강의는 type = 무료, amount = 0, quota = 무한대`
`ex)유료 강의는 type = 무료, amount = :inputAmount, quota = :inputAmount`

장점은 무료/유료 강의가 확실히 구분되므로 자신의 책임에 집중 할 수있고, 7개의 변수를 6개로 줄일 수 있습니다.
하지만 DB에서 조회한 값을 도메인에 넣는작업이 좀 더 복잡해 질 것 같습니다.   

번외로, 현재 quota(정원 수) 를 `Applicants`에서 관리하고 있는데 기존 수강생만 가지고 있던 일급 컬렉션을 해친다고 볼 수 있을 것 같습니다.  정원 수를 같이 관리해도 괜찮을까요?

ohtaeg commented:
> 혹시 자신의 고유 id의 경우도 따로 분리하기도 하나요? 여태까지 봤던 기억은 없어서요

네네 저도 실제로 실무에서  말씀해주신 공통적인 id, createdAt, updatedAt를 BaseEntitiy와 같이 만들어
상속을 받아 사용한 경험도 있습니다 :)

ohtaeg commented:
> 장점은 메인 도메인인 Session이 기존 7개에서 id, image, sessionInformation, applicants 4개만 변수로 가집니다.
하지만 결국 새로운 도메인 sessionInformation도 4개의 변수를 가집니다. 덩치를 일부 분리했지만 여전히 3개 이상의 많은 변수를 담고 있습니다.

3개 이상을 가지게 되었지만 최대한 분리하시려고 하셨기에 분리에 의의를 두어 지금 상태를 유지해주셔도 좋을 것 같아요 👍 👍 

> 하지만 DB에서 조회한 값을 도메인에 넣는작업이 좀 더 복잡해 질 것 같습니다.

네네 저도 강의를 아예 별도 도메인으로 서로 분리한다면 조금 더 복잡할 수도 있다고 생각되어지네요 :)
오히려 도메인으로 분리하지말고 강의 타입에 대해 인터페이스를 통해 다형성을 보장해주는 방법도 생각해볼 순 있을 것 같아요!
아직 미션들이 남아있으니 현재 상태를 유지하신 후 남은 미션을 진행하시면서 같이 고민해보아요 😄 

> 번외로, 현재 quota(정원 수) 를 Applicants에서 관리하고 있는데 기존 수강생만 가지고 있던 일급 컬렉션을 해친다고 볼 수 있을 것 같습니다. 정원 수를 같이 관리해도 괜찮을까요?

type, amount, quota를 하나의 객체로 잘 묶어주셨어요 💯 
저는 오히려 정원수는 강의 타입(유료, 무료)에 좀 더 밀접해보인다고 생각을 하기에
말씀해주신대로 일급 컬렉션을 해친다는 의견에 동의합니다 :)

ohtaeg commented on src/main/java/nextstep/courses/domain/course/session/Applicants.java:
```diff
@@ -0,0 +1,53 @@
+        return this.applicants.size();
+    }
+
+    public void addApplicant(NsUser applicant, Session.Type type) {
```
Session과 Applicants이 연관 관계이면서 Applicants이 Session을 직접적으로 의존하진 않지만
Session의 존재를 알게되는 나름(?) 양방향인 것 같아요 :)
단방향으로 바꿔보면 어떨까요? 😄 

[주셨던 번외  질문](https://github.com/next-step/java-lms/pull/340#issuecomment-1848293183)과 같이 고민해보시면 좋을 것 같아요!!

sunggyupaik replied:
네, 일단 `Session.Type`를 제거했습니다. 이 과정에서 유료와 무료를 구분하기 위해 메서드가 분리되었는데, 3단계에서 `Session`의 구조를 다시 변경하면서 개선해보겠습니다.

ohtaeg commented on src/main/java/nextstep/payments/domain/Payment.java:
```diff
@@ -26,4 +29,10 @@ public Payment(String id, Long sessionId, Long nsUserId, Long amount) {
+    public boolean isPaid(NsUser nsUser, Session session) {
+        return nsUser.isSame(this.nsUserId)
+                && session.sameId(this.sessionId)
+                && session.sameAmount(this.amount);
```
피드백 반영 👍 👍

ohtaeg reviewed (changes requested):
피드백 반영을 잘해주셨네요 👍 👍 
마지막으로 같이 고민해보면 좋을 것 같아 R.C 하였습니다!!
2단계는 이게 마지막 피드백이 될 것 같네요!!
화이팅입니다 🔥

sunggyupaik committed:
[refactor] 수강생과 강의의 의존성을 제거하라

강의와 수강생이 서로 의존성을 가지고 잇지만 강의가 수강생에 의존하는
것이 더 자연스러운 흐름이므로 수강생이 강의의 존재를 모르도록
개선합니다. 추후 일급 컬렉션을 보장하기 위해 정원 수를 분리해야 합니다

sunggyupaik committed:
[refactor] 과정의 공통 속성을 상속으로 개선하라

생성자, 생성일시, 수정일시를 공통 클래스로 분리합니다

sunggyupaik committed:
[refactor] 강의의 공통 속성을 별도 클래스로 분리하라

생성자, 생성일시, 수정일시를 별도의 클래스를 활용하여 상속바도록
개선합니다. 모든 클래스의 공통 속성으로 개선합니다

sunggyupaik commented:
피드백 반영 완료했습니다~

마감 날짜가 얼마 남지 않아서 짧은 주기로 계속 요청드렸는데 확인해주셔서 감사드립니다 :)   

`CreatorId`, `CreatedAt`, `UpdatedAt`을 가지는 `BaseEntity`로 공통 부분을 분리하였고 
`Session`은 현재 상태를 유지한 채로 다음 미션에서 필요 시에 다시 한번 구조를 개선해보도록 하겠습니다!

ohtaeg commented on src/main/java/nextstep/courses/domain/BaseEntity.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.time.LocalDateTime;
+
+public class BaseEntity {
```
👍 👍

ohtaeg reviewed (approved):
피드백 반영을 잘해주셨어요 👍 👍 
아직 여유가 있으니 지금 흐름대로  해주시면 좋을 것 같네요 :)
다음 미션도 화이팅입니다 🔥
